### PR TITLE
Completely revamp execute_in_process result API

### PR DESCRIPTION
**Pull request by dpeng817**
Completely revamps result API for`execute_in_process`. Removes a ton of extra cruft from implementation, and makes testing docs way simpler to read.

### CODE REVIEW

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
+from typing import Any, Dict, List, Optional
+
+from dagster import DagsterEvent, check
+from dagster.core.definitions import NodeDefinition, NodeHandle
+from dagster.core.definitions.utils import DEFAULT_OUTPUT
+from dagster.core.errors import DagsterInvariantViolationError
+from dagster.core.execution.plan.outputs import StepOutputHandle
+
+
+class InProcessResult:
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/execution/execute_in_process_result.py:10]: nit: `InProcessExecutionResult`

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+        self._output_capture = output_capture
+
+    @property
+    def success(self):
+        """bool: Whether execution was successful."""
+        return all([not step_event.is_step_failure for step_event in self.all_node_events])
+
+    @property
+    def all_node_events(self):
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/execution/execute_in_process_result.py:33]: is `node` useful? i feel like thats omissible here

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        self._handle = NodeHandle(node_def.name, parent=None)
+        self._event_list = all_events
+        self._run_id = run_id
+
+        self._output_capture = output_capture
+
+    @property
+    def success(self):
+        """bool: Whether execution was successful."""
+        return all([not step_event.is_step_failure for step_event in self.all_node_events])
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/execution/execute_in_process_result.py:30]: some rare cases of pipeline failure happen without step failures

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        # Get output from origin node
+        return _filter_outputs_by_handle(
+            self._output_capture, origin_handle, origin_output_def.name
+        )
+
+    def output_for_node(self, node_str, output_name=DEFAULT_OUTPUT):
+
+        # resolve handle of node that node_str is referring to
+        target_handle = NodeHandle.from_string(node_str)
+        target_node_def = self._node_def.ensure_graph_def().get_solid(target_handle).definition
```
Comment (alangenfeld) [inline python_modules/dagster/dagster/core/execution/execute_in_process_result.py:67]: should  _node_def just be a graph? It seems like we ensure graph in a few places

**Review by alangenfeld — COMMENTED** [review]
I don't remember any specific reasons for the current structure, I think it was just something we were trying. If the docs and tests point at this being clearer i think thats sufficient motivation.

**Review by sryza — COMMENTED** [review]
I like it

**Comment by sryza** [comment]
We'll need to add a section to the migration guide to cover these changes.  Also, how difficult would it be to support back-compat? At least for the non-composition cases?

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py (hunk 3 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/core/execution/execute_in_process_result.py:30]: Yea this made me uneasy. Gonna have to do something smarter here.

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py (hunk 4 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/core/execution/execute_in_process_result.py:67]: Realistically it should for now, but came from possibly targeting ops from jobs. Maybe putting the cart a little before the horse though.

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dpeng817) [reply python_modules/dagster/dagster/core/execution/execute_in_process_result.py:10]: totally

File: python_modules/dagster/dagster/core/execution/execute_in_process_result.py (hunk 2 above)
Comment (dpeng817) [reply python_modules/dagster/dagster/core/execution/execute_in_process_result.py:33]: Technically only step events, but I didnt want to use the name step lol

**Comment by dpeng817** [comment]
> We'll need to add a section to the migration guide to cover these changes. Also, how difficult would it be to support back-compat? At least for the non-composition cases?

I think backcompat is possible, but would be significant effort. In my mind, we could support backcompat by recursively constructing this result with whatever node def is described by `result_for_node`. This would mean keeping track of where you are in the descent stack, and adding additional cases to every method for dealing with ops. 

IMO it's a pretty straightforward migration from the old way to the new way, fwiw.

File: docs/content/guides/dagster/graph_job_op.mdx [hunk 5]
```diff
@@ -56,7 +56,14 @@ In both cases, you can load the pipeline / graph in Dagit with:
...
 
 ```python file=/guides/dagster/graph_job_op/execute_simple_graph.py startafter=start_execute endbefore=end_execute
-do_it_all.execute_in_process()
+result = do_it_all.execute_in_process()
+```
+
+A call to <PyObject module="dagster" object="GraphDefinition" method="execute_in_process" /> creates an <PyObject module="dagster" object="InProcessResult" /> object. Outputs can be retrieved using the <PyObject module="dagster" object="InProcessResult" method="output_for_node" /> method, and events can be retrieved using the <PyObject module="dagster" object="InProcessResult" method="events_for_node" /> method.
+
+```python file=/guides/dagster/graph_job_op/execute_simple_graph.py startafter=start_retrieve_output endbefore=end_retrieve_output
+output = result.output_for_node("do_something")
```
Comment (sryza) [inline docs/content/guides/dagster/graph_job_op.mdx:65]: You could make this a little bit clearer by including asserts that show what values you'd expect to come out.

**Review by sryza — APPROVED** [review]

**Pull request merged**